text/0000-conditional-compilation-checking.md

```diff
@@ -0,0 +1,486 @@
+user has selected, the target CPU architecture, the target OS, or other parameters under control
+of the user. Rust programs may use conditional compilation in these ways:
+
+* By applying the `#[cfg(c)]` attribute to language elements, where `c` is a condition.
+* By applying the `#[cfg_attr(c, attr)]` attribute to language elements, where `c` is a conditional
+  and `attr` is an attribute to apply.
+* By using the `cfg!(c)` built-in macro, where `c` is a condition. The compiler replaces the macro
+  call with a `true` or `false` literal.
+
+A _condition_ can take one of two forms:
+
+* A single identifier, such as `#[cfg(test)]` or `#[cfg(linux)]`. These are Boolean conditions;
+  they are either enabled or disabled.
+* A condition may test whether a given value is present in a named list of values. For example,
+  `#[cfg(feature = "lighting")]` tests whether the `lighting` feature is enabled. Note that a given
+  condition name may have any number of enabled values; for example, it is legal to invoke
+  `rustc --cfg feature="lighting" --cfg feature="bump_maps"`.
+
+## Checking conditions names
+
+`rustc` can optionally verify that all condition names used in source code are valid. "Valid" is
+distinct from "enabled". A _valid_ condition is one that is allowed to appear in source code; the
+condition may be enabled or disabled, but it is still valid. An _enabled_ condition is one which has
+been specified with a `--cfg foo` or `--cfg 'foo="value"'` option.
+
+For example, `rustc` can detect this bug, where the `test` condition is misspelled as `tset`:
+
+```rust
+if cfg!(tset) {    // uh oh, should have been 'test'
+   ...
+}
+```
+
+To catch this error, we give `rustc` the set of valid condition names:
+
+```bash
+rustc --cfg 'valid(name1, name2, ..., nameN)' ...
+```
+
+The `--cfg 'valid(...)'` option specifies the set of valid condition names.
+
+### Examples
+
+```bash
+rustc --cfg 'valid(cats, dogs, foo, bar)' ...
+```
+
+The `--cfg 'valid(...)'` option may be repeated. If it is repeated, then the condition names of
+all of the options are merged into a single set. For example:
+
+```bash
+rustc --cfg 'valid(cats)' --cfg 'valid(dogs)' ...
+```
+
+### Well-known condition names
+
+`rustc` defines a set of well-known conditions, such as `test`, `target_os`, etc. These conditions
+are always valid; it is not necessary to enable checking for these conditions. If these conditions 
+are specified in a `--cfg valid(...)` option, they will be ignored. This set of well-known names is
+a part of the stable interface of the compiler. New well-known conditions  may be added in the
+future, because adding a new name cannot break existing code. However, a name may not be removed
+from the set of well-known names, because doing would be a breaking change.
+
+These are the well-known conditions:
+
+* `feature`
+* `linux`
+* `test`
+* `target_os`
+* `target_arch`
+* `windows`
+* TODO: finish enumerating this list

```

Comment:
IMO you don't need to enumerate all of these, that can be left until implementation.

Reply:
Also, it's not `linux` - it's `unix`.

`"linux"` is [`target_os`](https://doc.rust-lang.org/reference/conditional-compilation.html?highlight=linux#target_os), `unix` is [`target_family`](https://doc.rust-lang.org/reference/conditional-compilation.html?highlight=linux#target_family). 

Only the latter is allowed as `#[cfg(unix)]`, the former is only as `#[cfg(target_os = "linux")], see [unix and windows](https://doc.rust-lang.org/reference/conditional-compilation.html?highlight=linux#unix-and-windows).